File: src/cmd/sync.rs
use std::mem::take;
use std::sync::Arc;

use anyhow::Result;
use async_trait::async_trait;
use clap::Args;

use crate::batch::{self, Task};
use crate::cmd::complete::{CompleteArg, CompleteCommand};
use crate::config::context::ConfigContext;
use crate::db::repo::Repository;
use crate::repo::current::get_current_repo_optional;
use crate::repo::ops::{RepoOperator, SyncResult};
use crate::repo::select::{RepoSelector, SelectManyReposOptions, SelectRepoArgs};
use crate::term::confirm::confirm_items;
use crate::{debug, outputln};

use super::Command;

/// Synchronize one or multiple repositories. Include cloning, pushing, pulling, etc.
#[derive(Debug, Args)]
pub struct SyncCommand {
    #[clap(flatten)]
    pub select_repo: SelectRepoArgs,

    /// By default, if you are currently in a repository, sync will synchronize the current
    /// repository; otherwise it will sync multiple repositories. With this option,
    /// multiple repositories will be synced regardless of your current location.
    #[arg(short)]
    pub recursive: bool,

    /// By default, when syncing multiple repositories, only repositories marked with sync
    /// flag will be synchronized. With this option, all repositories will be forcefully
    /// synced, ignoring the sync flag.
    #[arg(short)]
    pub force: bool,
}

#[async_trait]
impl Command for SyncCommand {
    fn name() -> &'static str {
        "sync"
    }

    async fn run(self, ctx: ConfigContext) -> Result<()> {
        debug!("[cmd] Run sync command: {:?}", self);

        if !self.recursive
            && let Some(repo) = get_current_repo_optional(&ctx)?
        {
            return self.sync_one(ctx, repo);
        }
        self.sync_many(ctx).await
    }

    fn complete() -> CompleteCommand {
        Self::default_complete()
            .args(SelectRepoArgs::complete())
            .arg(CompleteArg::new().short('r'))
            .arg(CompleteArg::new().short('f'))
    }
}

impl SyncCommand {
    async fn sync_many(self, mut ctx: ConfigContext) -> Result<()> {
        let selector = RepoSelector::new(&ctx, &self.select_repo);
        let mut opts = SelectManyReposOptions::default();
        if !self.force {
            opts.sync = Some(true);
        }
        let list = selector.select_many(opts)?;
        if list.items.is_empty() {
            outputln!("No repo to sync");
            return Ok(());
        }

        let mut names = list.display_names();
        confirm_items(&names, "Sync", "synchronization", "Repo", "Repos")?;

        let mut tasks = Vec::with_capacity(list.items.len());
        ctx.mute();
        let ctx = Arc::new(ctx);
        for (idx, repo) in list.items.into_iter().enumerate() {
            let task = SyncTask {
                name: Arc::new(take(&mut names[idx])),
                ctx: ctx.clone(),
                repo,
            };
            tasks.push(task);
        }

        let results = batch::run("Syncing", "Sync", tasks).await?;
        let results: Vec<String> = results
            .iter()
            .filter_map(|r| {
                let text = r.render(true);
                if text.is_empty() { None } else { Some(text) }
            })
            .collect();
        outputln!();
        if results.is_empty() {
            outputln!("No result to display");
            return Ok(());
        }

        for (idx, result) in results.iter().enumerate() {
            outputln!("{result}");
            if idx != results.len() - 1 {
                outputln!();
            }
        }

        Ok(())
    }

    fn sync_one(self, ctx: ConfigContext, repo: Repository) -> Result<()> {
        let op = RepoOperator::load(&ctx, &repo)?;
        let result = op.sync()?;
        let text = result.render(false);
        if text.is_empty() {
            outputln!("No result to display");
            return Ok(());
        }

        outputln!();
        outputln!("Result:");
        outputln!("{text}");
        Ok(())
    }
}

struct SyncTask {
    name: Arc<String>,
    ctx: Arc<ConfigContext>,
    repo: Repository,
}

#[async_trait]
impl Task<SyncResult> for SyncTask {
    fn name(&self) -> Arc<String> {
        self.name.clone()
    }

    async fn run(&self) -> Result<SyncResult> {
        RepoOperator::load(self.ctx.as_ref(), &self.repo)?.sync()
    }
}

File: cmd/sync.go
package cmd

import (
	"fmt"
	"os"

	"github.com/fioncat/roxide/pkg/batch"
	"github.com/fioncat/roxide/pkg/choice"
	"github.com/fioncat/roxide/pkg/context"
	"github.com/fioncat/roxide/pkg/db"
	"github.com/fioncat/roxide/pkg/repoutils"
	"github.com/fioncat/roxide/pkg/term"
	"github.com/spf13/cobra"
)

func NewSync() *cobra.Command {
	var opts syncOptions
	c := &cobra.Command{
		Use:   "sync [HEAD] [QUERY]",
		Short: "Sync multiple repositories or current repository (if you are in one)",

		Args: cobra.ArbitraryArgs,

		ValidArgsFunction: BuildCompletion(RepoCompletion),
	}

	c.Flags().BoolVarP(&opts.recursive, "recursive", "r", false, "force to sync multiple repositories")
	c.Flags().BoolVarP(&opts.force, "force", "f", false, "force sync, ignore sync flag")

	return Build(c, &opts)
}

type syncOptions struct {
	args []string

	recursive bool

	force bool
}

func (o *syncOptions) Complete(c *cobra.Command, args []string) error {
	o.args = args
	return nil
}

func (o *syncOptions) Run(ctx *context.Context) error {
	if !o.recursive {
		repo, err := repoutils.GetCurrentRepo(ctx)
		if err != nil {
			return err
		}
		if repo != nil {
			err = ctx.SetRepo(repo)
			if err != nil {
				return err
			}
			result, err := repoutils.Sync(ctx)
			if err != nil {
				return err
			}

			fmt.Fprintln(os.Stderr)
			display := result.Render(false)
			if display == "" {
				fmt.Fprintln(os.Stderr, "No result to display")
				return nil
			}

			fmt.Fprintln(os.Stderr, "Result:")
			fmt.Fprintln(os.Stderr, display)
			return err
		}
	}

	ch := choice.New(ctx, o.args)
	opts := choice.ManyOptions{}
	if !o.force {
		opts.Sync = db.BoolPtr(true)
	}
	list, err := ch.ManyLocal(opts)
	if err != nil {
		return err
	}

	if len(list.Items) == 0 {
		term.PrintInfo("No repo to sync")
		return nil
	}

	items := make([]string, 0, len(list.Items))
	for _, repo := range list.Items {
		items = append(items, repo.String())
	}

	err = term.ConfirmItems(items, "sync", "synchronization", "Repo", "Repos")
	if err != nil {
		return err
	}

	tasks := make([]*syncTask, 0, len(list.Items))
	for _, repo := range list.Items {
		repoCtx, err := ctx.Derive(repo)
		if err != nil {
			return err
		}
		tasks = append(tasks, &syncTask{ctx: repoCtx})
	}

	results, err := batch.Run("Sync", tasks)
	if err != nil {
		return err
	}

	displays := make([]string, 0)
	for _, result := range results {
		display := result.Render(true)
		if display != "" {
			displays = append(displays, display)
		}
	}

	fmt.Fprintln(os.Stderr)
	if len(displays) == 0 {
		fmt.Fprintln(os.Stderr, "No result to display")
		return nil
	}

	for i, display := range displays {
		fmt.Fprintln(os.Stderr, display)
		if i != len(displays)-1 {
			fmt.Fprintln(os.Stderr)
		}
	}

	return nil
}

type syncTask struct {
	ctx *context.Context
}

func (t *syncTask) Name() string {
	return t.ctx.GetRepo().String()
}

func (t *syncTask) Run() (*repoutils.SyncResult, error) {
	return repoutils.Sync(t.ctx)
}
